prometheus/go_collector_go117.go

```diff
@@ -146,22 +157,29 @@ func (c *goCollector) Collect(ch chan<- Metric) {
 			if v1 > v0 {
 				m.Add(unwrapScalarRMValue(sample.Value) - m.get())
 			}
-			m.Collect(ch)
 		case *gauge:
 			m.Set(unwrapScalarRMValue(sample.Value))
-			m.Collect(ch)
 		case *batchHistogram:
 			m.update(sample.Value.Float64Histogram(), c.exactSumFor(sample.Name))
-			m.Collect(ch)
 		default:
 			panic("unexpected metric type")
 		}
 	}
-
 	// ms is a dummy MemStats that we populate ourselves so that we can
 	// populate the old metrics from it.
 	var ms runtime.MemStats
 	memStatsFromRM(&ms, c.rmSampleMap)
+
+	c.rmSampleMu.Unlock()
+
+	// Export all the metrics to ch.
+	// At this point we must not access rmSampleBuf or rmSampleMap, because
+	// a concurrent caller could use it. It's safe to Collect all our Metrics,
+	// however, because they're updated in a thread-safe way while MemStats
+	// is local to this call of Collect.
```

Comment:
Hmm, I guess it's concurrency-safe as in "there are no low-level data races" but elements of `rmMetrics` are still updated (in lines 161, 163) so with a concurrent call to `Collect`, we might collect here a mix of metrics from different `metrics.Read` calls. Perhaps that's fine, but I very much doubt it.

Reply:
One thing that comes to mind is that at a given point in time, metrics that would otherwise be consistent with one another may no longer be. In most cases this is harmless, but I did also go out of my way to ensure that the memory metrics in `runtime/metrics` are kept consistent without a STW, so perhaps we should try to maintain that here. 😅 

Good catch. I'm inclined to fix this unless there are any objections.

Reply:
Please go ahead. Thanks @beorn7 for catching it and thanks @mknyszek for quick action.

Reply:
Sent #975.